### PR TITLE
Back up kubeconfig secret in GCP secret manager

### PR DESCRIPTION
**Pull request by chaodaiG**
This script overrides kubeconfig secret in cluster, and since kubernetes cluster doesn't backup secret, the old value could get lost. Utilize the versioning of GCP secret manager for backing up, so that old history won't get lost

### CODE REVIEW

File: gencred/merge_kubeconfig_secret.py [hunk 1]
```diff
@@ -34,7 +34,13 @@
 import tempfile
 import time
 
+# Import the Secret Manager client library.
+# Requires `pip install google-cloud-secret-manager`
+from google.cloud import secretmanager
+
 reAutoKey = re.compile('^config-(\\d{8})$')
+gcp_secret_key = "prowkubeconfigbackup"
```
Comment (michelle192837) [inline gencred/merge_kubeconfig_secret.py:42]: nit: Should this be ALL_CAPS?

**Comment by chaodaiG** [comment]
POC
Screen Shot 2020-12-10 at 4 31 04 PM

Screen Shot 2020-12-10 at 4 31 21 PM

File: gencred/merge_kubeconfig_secret.py [hunk 2]
```diff
@@ -44,11 +50,31 @@ def call(cmd, **kwargs):
...
+        timeout=10,  # seconds
         universal_newlines=True,
         **kwargs,
     )
 
+
+def save_secret_to_gcp(project_id, data, secret_id=gcp_secret_key):
+    dst = f"projects/{project_id}/secrets/{secret_id}"
+    print('Saving secret to %s' % dst)
+    client = secretmanager.SecretManagerServiceClient()
```
Comment (michelle192837) [inline gencred/merge_kubeconfig_secret.py:62]: Huh, no auth needed?

File: gencred/merge_kubeconfig_secret.py [hunk 3]
```diff
@@ -114,12 +151,18 @@ def validateArgs(args):
 
 
 if __name__ == '__main__':
-    parser = argparse.ArgumentParser(description='Merges the provided kubeconfig file into a kubeconfig file living in a kubernetes secret in order to add new cluster contexts to the secret. Requires kubectl and base64.') #pylint: disable=line-too-long
+    parser = argparse.ArgumentParser(
+        description='Merges the provided kubeconfig file into a kubeconfig file living in a kubernetes secret in order to add new cluster contexts to the secret. Requires kubectl and base64.')  # pylint: disable=line-too-long
```
Comment (michelle192837) [inline gencred/merge_kubeconfig_secret.py:155]: This should be updated to note the new google-cloud-secret-manager requirement.

Comment (chaodaiG) [reply gencred/merge_kubeconfig_secret.py]: This is good to know, updated

File: gencred/merge_kubeconfig_secret.py (hunk 2 above)
Comment (chaodaiG) [reply gencred/merge_kubeconfig_secret.py:66]: Actually it's needed, it requires ADC auth which is done by gcloud, I have added the required command in the description below, as well as `requirement` section on top

File: gencred/merge_kubeconfig_secret.py (hunk 1 above)
Comment (chaodaiG) [reply gencred/merge_kubeconfig_secret.py]: Done, to be more careful maybe they should be `_` prefixed, but I'm not imagining anyone importing this file, so let's keep them as they are

**Comment by chaodaiG** [comment]
Something is not right in the test, it failed with:
```
Collecting google-cloud-secret-manager==2.1.0 (from -r /home/prow/go/src/github.com/kubernetes/test-infra/requirements3.txt (line 13))
 (  Could not find a version that satisfies the requirement google-cloud-secret-manager==2.1.0 (from -r /home/prow/go/src/github.com/kubernetes/test-infra/requirements3.txt (line 13)) (from versions: 0.1.0, 0.1.1, 0.2.0, 1.0.0)
```

But this was added as python3 requirements, which 2.1.0 is [officially supported](https://pypi.org/project/google-cloud-secret-manager/) version. The official document also says `The last version of this library compatible with Python 2.7 is google-cloud-secret-manager==1.0.0.`,  the error message `from versions: 0.1.0, 0.1.1, 0.2.0, 1.0.0` indicates that the test was fetching python3 dependencies using python2.

File: gencred/BUILD.bazel [hunk 4]
```diff
@@ -34,3 +35,12 @@ go_binary(
     embed = [":go_default_library"],
     visibility = ["//visibility:public"],
 )
+
+py_binary(
+    name = "merge_kubeconfig_secret",
+    srcs = ["merge_kubeconfig_secret.py"],
+    python_version = "PY3",
+    deps = [
+        requirement("google-cloud-secret-manager"),
```
Comment (fejta) [inline gencred/BUILD.bazel:44]: google-cloud-secret-manager depends on asteroid, thus you need to add `requirement('asteroid`)` in here (as well as any other dependencies

Comment (chaodaiG) [reply gencred/BUILD.bazel:44]: Had an offline discussion, the error message with `asteroid` was probably a misleading error, which very likely was a truncated error message meant to print https://github.com/kubernetes/test-infra/blob/master/requirements3.txt but only got to the first line.

**Comment by chaodaiG** [comment]
No clue how to resolve the bazel test error, will park this PR for here for now

**Comment by fejta** [comment]
My three thoughts are:
1) Golang is generally going to be a better choice around k8s than python
2) Despite the error message, I suspect a missing transitive dependency in the requirements -- either in requirements.txt or more likely the require() build rule
3) Perhaps there's a stale version of the pip project list

**Pull request closed without merging**